### ISSUE TITLE
Refactor FireFox extension to use WebExtensions

### ISSUE THREAD

**Issue by SpaceK33z** [facebook/react-devtools#524]
FireFox now supports the [WebExtensions API](https://developer.mozilla.org/en-US/Add-ons/WebExtensions), which is almost exactly like the Chrome extension API.

Mozilla is planning to drop support for the existing extension API by the end of 2017 ([source](https://blog.mozilla.org/addons/2016/11/23/add-ons-in-2017/)).

So the FF extension will need to be rebuild to be compatible with the WebExtensions API before the end of 2017. However, this is good news, because the current Chrome version will perhaps work out-of-the-box and if not, it would need only small changes. So it shouldn't be too much work.

**Comment by zalmoxisus** [comment]
`devtools.panels` are not supported yet. See http://arewewebextensionsyet.com/#devtools.panels and https://bugzilla.mozilla.org/show_bug.cgi?id=1300587. [Some progress has been made recently](https://reviewboard.mozilla.org/r/76818/diff/13/), but I wasn't able to create a panel in FF Nightly yet.

**Comment by SpaceK33z** [comment]
@zalmoxisus ah, that's a shame. Still might be worth to keep this open to monitor this.

**Comment by SpaceK33z** [comment]
If there are questions about the WebExtensions API or missing features, feel free to ask @rob--w. He's a colleague of mine and contributed much to the WebExtensions API as Mozilla volunteer.

**Comment by clarkbw** [comment]
[bug 1300587](https://bugzilla.mozilla.org/show_bug.cgi?id=1300587) has been fixed and is on target for Firefox 54 which goes live to release on June 13th.  Now is probably a good time to get started on this work.  Let me know how I can help.

**Comment by clarkbw** [comment]
I was checking this out today on Firefox 54 (DevEdition) and things mostly work https://twitter.com/clarkbw/status/857979112119349249

However I'm not sure what would be the best way to proceed. 

@gaearon do you have any thoughts?

I'm able to run the Chrome extension as is in Firefox and it would likely take some modifications (still need to investigate them all) to make sure things look normal and work correctly.  I'm wondering about how we handle any differences continuing down the line.

We could:

A shared source directory like `webextensions` for common code and then the `chrome` and `firefox` directories remain as light wrappers?

Or (my preference) converge on the `chrome` directory and do some of the fixes directly in there.  This hopefully leads to less divergence.  We could possibly rename that directory to `webextension` since it does also work in Opera but it doesn't matter too much to me.

Or some other option...

**Comment by gaearon** [comment]
What you suggest seems sensible. Need to know the impact of the changes to tell for sure.

**Comment by clarkbw** [comment]
Writing out some notes here, I'll probably just keep updating this comment.

### manifest.json

Need to modify the `manifest.json`, possibly at the build step to include this for Firefox only

```json
  "applications": {
    "gecko": {
      "strict_min_version": "54"
    }
  }
```

The [application](https://developer.mozilla.org/en-US/Add-ons/WebExtensions/manifest.json/applications) field will work in Chrome but Chrome will complain about it being in the manifest so I'd rather not build it into a shared manifest if I can.

**Comment by clarkbw** [comment]
Really small amount of changes other than what is required above.  Here's my branch so far: https://github.com/facebook/react-devtools/compare/master...clarkbw:firefox-webextension

**Comment by gaearon** [comment]
That's awesome! Want to send a PR?

**Comment by clarkbw** [comment]
> That's awesome! Want to send a PR?

I've got one bug with left with... flexbox.  When something in the Inspector sizes beyond the available space it pushes the sidebar and window out instead of running underneath the sidebar.  I don't see this behavior when I run it in a tab, only when I'm running it in a panel.

I'm looking at the differences between the [Firefox panel CSS](https://github.com/facebook/react-devtools/blob/master/shells/firefox/data/panel.html) and the [Chrome panel CSS](https://github.com/facebook/react-devtools/blob/master/shells/chrome/panel.html) to understand what's wrong.

**Comment by clarkbw** [comment]
To get the `manifest.json` pieces working here are some options:

1. Land the [required changes to the `manifest.json`](https://github.com/facebook/react-devtools/compare/master...clarkbw:firefox-webextension-manifest) file and add notes to the README that describe the Chrome warning people will see when developing the extension in Chrome (I don't believe you would see this as a regular extension user).

<img width="768" alt="screen shot 2017-05-01 at 12 00 20 pm" src="https://cloud.githubusercontent.com/assets/2134/25590567/efb3b994-2e65-11e7-818f-ced26dc4309b.png">

2. I could wait until #632 has landed.  If we are generating a `manifest.json` and automating the building of the add-on I could just generate a second add-on with an alternate manifest and save that to`build/react-devtools-firefox` (current proposal saves the existing bundle to `build/react-devtools-chrome`)

**Comment by gaearon** [comment]
I probably wouldn't *wait* for #632 to land, but I would make it so that we still have two folders.

**Comment by bvaughn** [comment]
Closed via #802
